### PR TITLE
Complete the server side requirements for EDTR Prototype

### PR DESCRIPTION
**Pull request by manzoorwanijk**
Completes #1855 #1856 #1857 

## Checklist

* [x] I have read the documentation relating to systems affected by this pull request, see https://github.com/eventespresso/event-espresso-core/tree/master/docs
* [x] User input is adequately validated and sanitized
* [x] all publicly displayed strings are internationalized (usually using `esc_html__()`, see https://codex.wordpress.org/I18n_for_WordPress_Developers)
* [ ] My code is tested.
* [x] My code follows the Event Espresso code style.
* [x] My code has proper inline documentation.
* [ ] My code accounts for when the site is in Maintenance Mode (MM2 especially disallows usage of models)

### CODE REVIEW

File: core/domain/services/graphql/connections/RootQueryDatetimesConnection.php [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+     * @param $info
+     * @return array
+     * @throws Exception
+     * @since $VID:$
+     */
+    public function resolveConnection($entity, $args, $context, $info)
+    {
+        $resolver = new DatetimeConnectionResolver($entity, $args, $context, $info);
+        return $resolver->get_connection();
+    }
```
Comment (tn3rb) [inline core/domain/services/graphql/connections/RootQueryDatetimesConnection.php]: this is also almost the same logic as `core/domain/services/graphql/connections/RootQueryTicketsConnection.php` and could likely be extracted to the parent class, but utilize a `getConnectionResolver()` method for returning the correct resolver to call `get_connection()` on.

File: core/domain/services/graphql/connections/RootQueryDatetimesConnection.php [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+     * @param $context
+     * @param $info
+     * @return EE_Base_Class
+     * @since $VID:$
+     */
+    public function resolveNode($id, $args, $context, $info)
+    {
+        return $this->model->get_one_by_ID($id);
+    }
+}
```
Comment (tn3rb) [inline core/domain/services/graphql/connections/RootQueryDatetimesConnection.php]: also identical to logic in `core/domain/services/graphql/connections/RootQueryTicketsConnection.php` and therefore can be extracted

File: core/domain/services/graphql/connection_resolvers/DatetimeConnectionResolver.php [hunk 3]
```diff
@@ -131,9 +132,52 @@ public function get_query_args()
...
+         */
+        if (! empty($this->args['where']['orderby']) && is_array($this->args['where']['orderby'])) {
+            $query_args['order_by'] = [];
+            foreach ($this->args['where']['orderby'] as $orderby_input) {
+                $query_args['order_by'][ $orderby_input['field'] ] = $orderby_input['order'];
+            }
+        } elseif ($offset) {
+            $compare = ! empty($last) ? '<' : '>';
+            $where_params['DTT_ID'] = array($compare, $offset);
+        }
```
Comment (tn3rb) [inline core/domain/services/graphql/connection_resolvers/DatetimeConnectionResolver.php]: nearly identical to lines 134 - 142 of `core/domain/services/graphql/connection_resolvers/TicketConnectionResolver.php`
Can be extracted and moved into `AbstractConnectionResolver`

File: core/domain/services/graphql/data/mutations/DatetimeMutation.php [hunk 4]
```diff
@@ -34,8 +34,40 @@ public static function prepareFields(array $input)
...
+        // Remove all the existing related tickets
+        $entity->_remove_relations($relationName);
+
+        foreach ($tickets as $ID) {
+            $entity->_add_relation_to(
+                $ID,
+                $relationName
+            );
+        }
+    }
```
Comment (tn3rb) [inline core/domain/services/graphql/data/mutations/DatetimeMutation.php]: more nearly identical code that can be abstracted and extracted to the parent

**Comment by tn3rb** [comment]
@manzoorwanijk 

I took the liberty to refactor things and push some of the common code into new abstract parent classes. Hope you don't mind!
I'm going to approve the PR for the work that you did, but plz look over the changes I made to ensure that I didn't mess something up.

**Comment by manzoorwanijk** [comment]
@tn3rb Thank you for those changes. There was that one change needed to avoid the abstract class being loaded into connection collection.

I know there is some more such refactoring needed as we move forward, but that would be done when we get the things working properly for EDTR.

**Pull request merged**